### PR TITLE
overlay: add overlay support to parts (CRAFT-470)

### PR DESCRIPTION
**Pull request by cmatsuoka**
Define the overlay work directory. Add part properties to retrieve
overlay-related subdirectories and identify if a part has overlay
parameters or overlay visibility.

Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

### CODE REVIEW

**Review by cjp256 — APPROVED** [review]
LGTM

File: craft_parts/dirs.py [hunk 1]
```diff
@@ -35,6 +35,26 @@ def work_dir(self) -> Path:
         """Return the root of the work directories used for project processing."""
         return self._work_dir
 
+    @property
+    def overlay_dir(self) -> Path:
+        """Return the directory containing work subdirectories for overlays."""
+        return self._work_dir / "overlay"
```
Comment (facundobatista) [inline craft_parts/dirs.py:41]: Do these need to be properties?

Why not keep them simple, and just, in `__init__`, do `self.overlay_dir = self._work_dir / "overlay"` ?

**Review by facundobatista — CHANGES_REQUESTED** [review]
General idea looks great, but I'd go for a more simple implementation if possible.

File: craft_parts/dirs.py (hunk 1 above)
Comment (cmatsuoka) [reply craft_parts/dirs.py:41]: `ProjectDirs` is minimally exposed in the codebase so we may want to change all its properties to attributes.

Comment (cmatsuoka) [reply craft_parts/dirs.py]: Done.

**Comment by cmatsuoka** [comment]
`Final` typing hint not available in python 3.7, so not using it for now.

**Review by sergiusens — APPROVED** [review]

**Review by facundobatista — APPROVED** [review]
Great, thanks!

**Pull request merged**